mungegithub/mungers/submit-queue-batch.go

```diff
@@ -0,0 +1,282 @@
+	sq.Lock()
+	defer sq.Unlock()
+	for _, pull := range batch.Pulls {
+		if _, ok := sq.githubE2EQueue[pull.Number]; ok {
+			return true
+		}
+	}
+	return false
+}
+
+// matchesCommit determines if the batch can be merged given some commits.
+// That is, does it contain exactly:
+// 1) the batch's BaseSha
+// 2) (optional) merge commits for PRs in the batch
+// 3) any merged PRs in the batch are sequential from the beginning
+// The return value is -1 on error, else the number of PRs already merged.
+func (batch *Batch) matchesCommits(commits []*githubapi.RepositoryCommit) int {
+	if len(commits) == 0 {
+		return -1
+	}
+
+	shaToPR := make(map[string]int)
+
+	for _, pull := range batch.Pulls {
+		shaToPR[pull.Sha] = pull.Number
+	}
+
+	matchedPRs := []int{}
+
+	// convert the list of commits into a DAG for easy following
+	dag := make(map[string]*githubapi.RepositoryCommit)
+	for _, commit := range commits {
+		dag[*commit.SHA] = commit
+	}
+
+	ref := *commits[0].SHA
+	for {
+		if ref == batch.BaseSha {
+			break // found the base ref (condition #1)
+		}
+		commit, ok := dag[ref]
+		if !ok {
+			return -1 // ref not in partial list of commits
+		}
+		if len(commit.Parents) == 2 && commit.Message != nil && (*commit.Message)[:5] == "Merge" {
+			// looks like a merge commit!
+
+			// first parent is the normal branch
+			ref = *commit.Parents[0].SHA
+			// second parent is the PR
+			pr, ok := shaToPR[*commit.Parents[1].SHA]
+			if !ok {
+				return -1 // merge of a PR not in batch
+			}
+			matchedPRs = append(matchedPRs, pr)
+		} else {
+			return -1 // unhandled non-merge commit
+		}
+	}
+
+	// Now, ensure that the merged PRs are ordered correctly.
+	for i, pr := range matchedPRs {
+		if batch.Pulls[len(matchedPRs)-1-i].Number != pr {
+			return -1
+		}
+	}
+	return len(matchedPRs)
+}
+
+// batchIsApplicable returns whether a successful batch result can be used--
+// 1) some of the batch is still unmerged and in the queue.
+// 2) the recent commits are the batch head ref or merges of batch PRs.
+// 3) all unmerged PRs in the batch are still in the queue.
+// The return value is -1 on error, else the number of PRs already merged.
+func (sq *SubmitQueue) batchIsApplicable(batch Batch) int {
+	// batch must intersect the queue
+	if !sq.batchIntersectsQueue(batch) {
+		return -1
+	}
+	commits, err := sq.githubConfig.GetBranchCommits(batch.BaseName)
+	if err != nil {
+		glog.Errorf("Error getting commits for batchIsApplicable: %v", err)
+		return -1
+	}
+	return batch.matchesCommits(commits)
+}
+
+func (sq *SubmitQueue) handleGithubE2EBatchMerge() {
+	repo := sq.githubConfig.Org + "/" + sq.githubConfig.Project
+	for range time.Tick(1 * time.Minute) {
+		jobs, err := getSuccessfulBatchJobs(repo, sq.BatchURL)
+		if err != nil {
+			glog.Errorf("Error reading batch jobs from Prow URL %v", sq.BatchURL)
+			continue
+		}
+		batches := sq.getCompleteBatches(jobs)
+		for _, batch := range batches {
+			if sq.batchIsApplicable(batch) < 0 {
```

Comment:
You run this condition here, and directly in the following function. Could we do it once?